pkg/stanza/fileconsumer/internal/trie/trie_test.go

```diff
@@ -0,0 +1,561 @@
+					value:         []byte("ABCD"),
+					matchExpected: true,
+				},
+				{
+					value:         []byte("ABCDE"),
+					matchExpected: true,
+				},
+			},
+		},
+		{
+			name:         "TrieCase_DeleteFalse",
+			initialItems: []string{"ABCDEFG"},
+			testCases: []testCase{
+				{
+					value:          []byte("ABCDEFG"),
+					delete:         true,
+					deleteExpected: true,
+				},
+				{
+					value: []byte("ABCD"),
+				},
+				{
+					value:  []byte("XYZ"),
+					delete: true,
+					// it should be false, as we haven't inserted such values
+					deleteExpected: false,
+				},
+			},
+		},
+		{
+			name:         "TrieCase_Complex",
+			initialItems: []string{"ABCDE", "ABC"},
+			testCases: []testCase{
+				{
+					value:         []byte("ABCDEXYZ"),
+					matchExpected: true,
+				},
+				{
+					value:         []byte("ABCXYZ"),
+					matchExpected: true,
+				},
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		trie := NewTrie()
+		for _, k := range tc.initialItems {
+			trie.Put([]byte(k))
+		}
+		t.Run(tc.name, func(t *testing.T) {
+			for _, T := range tc.testCases {
+				runTest(t, trie, &T)
+			}
+		})
+	}
+}
+
+func runTest(t *testing.T, trie *Trie, tc *testCase) {
+	if tc.delete {
+		// Delete the value and check if it was deleted successfully
+		assert.Equal(t, trie.Delete(tc.value), tc.deleteExpected)
+	} else {
+		assert.Equal(t, trie.HasKey(tc.value), tc.matchExpected)
+		if !tc.matchExpected {
+			trie.Put(tc.value)
+		}
+	}
+}
+
+func TestTrieOpSequences(t *testing.T) {
+	opTree{
+		continuations: map[string]opTree{
+			"Found:ABC": opTree{ // First poll finds only ABC.
+				ops: []testOp{
+					put("ABC"),
+				},
+				continuations: map[string]opTree{
+					"Done:ABC": opTree{ // Finish reading ABC and remove from trie
+						ops: []testOp{
+							del("ABC", true, "was just added"),
+							has("ABC", false, "was just deleted"),
+						},
+					},
+					"Found:ABCDEF": opTree{ // Next poll finds ABCDEF
+						ops: []testOp{
+							has("ABCDEF", true, "recognize ABC w/ DEF appended"),
+							put("ABCDEF"), // TODO HasKey returns true, so how do we know to call this?
+							has("ABC", false, "should push ABC down to ABCDEF"),
+						},
+						// TODO When we started reading this file, we identified it as ABC.
+						//      However, we have an updated understanding that it is ABCDEF.
+						//      Should we only have to delete ABC, or only ABCDEF, or both?
+						continuations: map[string]opTree{
+							"DeleteAs:ABC": opTree{ // Done reading the file, remove it as ABC
+								ops: []testOp{
+									del("ABC", false, "should have been pushed down when ABCDEF was added"),
+									has("ABCDEF", true, "should not have been deleted"),
+								},
+								continuations: map[string]opTree{
+									"DeleteAs:ABCDEF": opTree{ // Also remove it as ABCDEF
+										ops: []testOp{
+											del("ABCDEF", true, "just confirmed it exists"), // TODO this fails to delete
+										},
+									},
+								},
+							},
```

Comment:
I agree w/ your assessment [here](https://github.com/VihasMakwana/opentelemetry-collector-contrib/pull/2#discussion_r1278590622) that we only need to delete `ABCDEF`, so we can remove this case.

The main thing here is that we need to know what to pass to Delete. I think my suggestion of switching back to `Get` could work here because we can save a value on the reader that indicates what to pass to Delete when the time comes.